crates/bevy_ecs/src/query/state.rs

```diff
@@ -138,6 +138,55 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
         state
     }
 
+    /// Transform a query into a more generic query. If the original query state did
+    /// not have the same access this will panic.
+    /// Probably should not call update archetype generation on this query state as the results will
+    /// be very unpredictable as new archetypes could be added that don't match old query
+    pub fn generalize<NewQ: WorldQuery>(&self, world: UnsafeWorldCell) -> QueryState<NewQ, ()> {
+        if !Q::IS_ARCHETYPAL || !F::IS_ARCHETYPAL || !NewQ::IS_ARCHETYPAL {
+            panic!("generalizing is not allow with queries that use `Changed` or `Added`");
+        }
+        // SAFETY: we are using unsafe world to access &Components which should always be safe
+        let fetch_state = unsafe { NewQ::new_state(world.world().components()) };
```

Comment:
```suggestion
        let fetch_state = NewQ::new_state(world.components());
```

`UnsafeWorldCell` has safe accessors for metadata like `Components`